### PR TITLE
PXC-3430, PXC-3433: Merge PS 8.0.21 & Galera 4.5

### PR DESCRIPTION
**Pull request by dutow**

### CODE REVIEW

File: mysql-test/mysql-test-run.pl [hunk 1]
```diff
@@ -337,10 +309,14 @@ sub set_term_args {
   interactive_utilities
   jp
   stress
+  galera
+  galera_3nodes
+  galera_sr
+  galera_3nodes
```
Comment (kennt-percona) [inline mysql-test/mysql-test-run.pl]: looks like "sr" was put on a separate line (rather than galera_3nodes_sr)

File: sql/wsrep_mysqld.cc [hunk 2]
```diff
@@ -1044,7 +1044,8 @@ int wsrep_init() {
     }
 
     char ssl_opts[4096];
-    SslAcceptorContext::populate_wsrep_ssl_options(ssl_opts, sizeof(ssl_opts));
+    // TODO
```
Comment (kennt-percona) [inline sql/wsrep_mysqld.cc]: Can this be removed? or is there something left to do?

File: sql/wsrep_thd.cc [hunk 3]
```diff
@@ -348,8 +348,13 @@ int wsrep_abort_thd(const THD *bf_thd, THD *victim_thd, bool signal) {
 }
 
 bool wsrep_bf_abort(const THD *bf_thd, THD *victim_thd) {
+  DBUG_TRACE;
   WSREP_LOG_THD(const_cast<THD *>(bf_thd), "BF aborter before");
+
+  //wsrep_lock_thd_query(victim_thd, true);
```
Comment (kennt-percona) [inline sql/wsrep_thd.cc]: Can these commented out lines be removed?

File: storage/innobase/lock/lock0lock.cc [hunk 4]
```diff
@@ -902,22 +902,27 @@ static void wsrep_kill_victim(const trx_t *const trx, const lock_t *lock) {
         } else {
           ib::info() << "*** Victim TRANSACTION:";
         }
-        wsrep_trx_print_locking(stderr, lock->trx, 3000);
+        wsrep_trx_print_locking(stderr, victim_trx, 3000);
         ib::info() << "*** WAITING FOR THIS LOCK TO BE GRANTED:";
 
-        if (lock_get_type(lock) == LOCK_REC) {
-          lock_rec_print(stderr, lock);
+#ifdef OUT
```
Comment (kennt-percona) [inline storage/innobase/lock/lock0lock.cc]: can this block of code be removed?

File: storage/innobase/lock/lock0lock.cc [hunk 5]
```diff
@@ -3822,7 +3840,7 @@ const lock_t *lock_table_other_has_incompatible(
        lock = UT_LIST_GET_PREV(tab_lock.locks, lock)) {
     if (lock->trx != trx && !lock_mode_compatible(lock_get_mode(lock), mode) &&
         (wait || !lock_get_wait(lock))) {
-#ifdef WITH_WSREP
+#ifdef WITH_WSREP_OUT
```
Comment (kennt-percona) [inline storage/innobase/lock/lock0lock.cc]: Can this block of code just be removed?

File: storage/innobase/lock/lock0lock.cc [hunk 6]
```diff
@@ -6547,6 +6551,18 @@ void lock_trx_release_locks(trx_t *trx) /*!< in/out: transaction */
 
   trx_mutex_enter(trx);
 
+#ifdef WITH_WSREP_OUT
```
Comment (kennt-percona) [inline storage/innobase/lock/lock0lock.cc]: similar to above, can this block be removed?

File: mysql-test/collections/disabled.def [hunk 7]
```diff
@@ -221,6 +221,11 @@ galera.galera_sst_rsync : BUG#0 rsync feature removed
 galera.galera_ist_rsync : BUG#0 rsync feature removed
 galera.galera_truncate : BUG#0 result depends on which node indexing
 
+galera.high_prio_trx_1 : BUG#0 Failing after 8.0.21 merge
```
Comment (kamil-holubicki) [inline mysql-test/collections/disabled.def]: Should we leave them disabled? Or at least provide the reason they are failing and why it is OK to disable these tests.

File: storage/innobase/handler/ha_innodb.cc [hunk 8]
```diff
@@ -1928,31 +1928,22 @@ ulint thd_start_time_in_secs(THD *thd) /*!< in: thread handle, or NULL */
 
 /** Enter InnoDB engine after checking the max number of user threads
 allowed, else the thread is put into sleep.
-<<<<<<< HEAD
 @param[in,out]	prebuilt	row prebuilt handler */
-static inline void innobase_srv_conc_enter_innodb(row_prebuilt_t *prebuilt) {
+static inline dberr_t innobase_srv_conc_enter_innodb(row_prebuilt_t *prebuilt) {
```
Comment (kamil-holubicki) [inline storage/innobase/handler/ha_innodb.cc]: It looks like
@return InnoDB error code. */
is missing in comment

File: storage/innobase/lock/lock0lock.cc [hunk 9]
```diff
@@ -5424,7 +5436,7 @@ static void rec_queue_validate_latched(const buf_block_t *block,
                                impl_trx)) {
           ib::info() << "WSREP impl BF lock conflict";
         }
-        ut_a(lock_get_wait(other_lock));
+        //ut_a(lock_get_wait(other_lock));
```
Comment (kamil-holubicki) [inline storage/innobase/lock/lock0lock.cc]: remove?

**Review by kamil-holubicki — COMMENTED** [review]
Some minor comments.
+ Suggestion: Probably too late right now, but in the future, I would insist on having separate commits for resolving conflicts, fixing mtr tests, etc... At least 4 commits here instead of 1. It would make review easier (we always can squash after all, but to be honest I'm not the big fan of squashing unrelated things)

File: mysql-test/collections/disabled.def (hunk 7 above)
Comment (dutow) [reply mysql-test/collections/disabled.def]: I've created a separate jira issue about these.

**Comment by dutow** [comment]
@kamil-holubicki There are 2 commits here, 1 for the 8.0.21 merge, and one for the galera 4.5 update. I don't see much point in separating the test and code conflicts into different commits (they are in different locations, easily viewable even on github). And for the code changes themselves, separating those into more conflicts isn't trivial - while there are some non trivial changes there, that's part of the conflict resolution, I wouldn't be 100% sure which parts of it to move into different commits. Keeping it together is simpler and cleaner.

File: sql/wsrep_mysqld.cc (hunk 2 above)
Comment (dutow) [reply sql/wsrep_mysqld.cc]: Yes, removed.

File: sql/wsrep_thd.cc (hunk 3 above)
Comment (dutow) [reply sql/wsrep_thd.cc]: Yes, removed.

File: storage/innobase/lock/lock0lock.cc (hunk 4 above)
Comment (dutow) [reply storage/innobase/lock/lock0lock.cc]: yes, removed.

File: storage/innobase/lock/lock0lock.cc (hunk 5 above)
Comment (dutow) [reply storage/innobase/lock/lock0lock.cc]: yes, removed.

File: storage/innobase/lock/lock0lock.cc (hunk 6 above)
Comment (dutow) [reply storage/innobase/lock/lock0lock.cc]: yes, removed.

File: storage/innobase/lock/lock0lock.cc (hunk 9 above)
Comment (dutow) [reply storage/innobase/lock/lock0lock.cc]: removed.

File: storage/innobase/handler/ha_innodb.cc (hunk 8 above)
Comment (dutow) [reply storage/innobase/handler/ha_innodb.cc]: commend added.

**Review by kamil-holubicki — APPROVED** [review]
lgtm.
We just need to be aware that we have the issue related to disabled MTR tests which was not there before.

**Review by kennt-percona — APPROVED** [review]
lgtm

**Comment by kamil-holubicki** [comment]
@dutow By splitting I meant having separate commits for these (of course if possible):
Also:
* Fixing test failures
* Porting new BF abort code from codership sources
* Refactoring SSL code to the new SSL API

**Review by kamil-holubicki — APPROVED** [review]

**Comment by dutow** [comment]
https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/128/

CentOS 7 debug fails because of PXC-3490,which is a jenkins issue. Otherwise same result as before.

**Pull request merged**